llmfoundry/models/layers/ffn.py

```diff
@@ -18,6 +20,47 @@
 
 log = logging.getLogger(__name__)
 
+_FFN_ACT_FN_DEFAULT = {
+    'name': 'gelu',
+    'approximate': 'none',
+}
+
+
+def resolve_ffn_act_fn(
+    config: Optional[dict] = None,) -> Callable[[torch.Tensor], torch.Tensor]:
+    """Resolve the activation function for the feed-forward network.
+
+    Args:
+        config (Optional[dict]): The configuration dictionary for the activation function.
+            The dict config must specify a 'function' or the 'name' of a function.
+            If 'function' is specified, a Callable function is expected. If 'name' is
+            specified, the name is expected to be the name of a `torch.nn.functional`
+            function. All of other key values pairs are bound to the function as a partial.
+
+    Returns:
+        Callable[[torch.Tensor], torch.Tensor]: The activation function.
+    """
+    config = deepcopy(config or _FFN_ACT_FN_DEFAULT)
```

Comment:
```suggestion
    if config is None:
        return _DEFAULT_ACT_FN
    config = deepcopy(config or _FFN_ACT_FN_DEFAULT)
```

Reply:
circular dependency
generating _DEFAULT_ACT_FN uses this fn
https://github.com/mosaicml/llm-foundry/pull/805/files/96b6330b8c3b51294ef041f7c3110c48ca655e46#diff-df5fd070e28295f023dcec5a8ad79df3cfd90c6d7a141f9714164f8ceff0fc93R62

Reply:
I can do 
```suggestion
    if config is None:
        config = _FFN_ACT_FN_DEFAULT
    config = deepcopy(config)
```

Reply:
ah, I see. How about remove the optional part, so the function accepts `config: dict`. It looks like `resolve_ffn_act_fn` is never called on `None` or without args. The default is already nicely explicitly set in other places with `_DEFAULT_ACT_FN` (like in `resolve_hidden_size`)

Reply:
[default here is None](https://github.com/mosaicml/llm-foundry/pull/805/files/96b6330b8c3b51294ef041f7c3110c48ca655e46#diff-df5fd070e28295f023dcec5a8ad79df3cfd90c6d7a141f9714164f8ceff0fc93R177)
gets passed in [here](https://github.com/mosaicml/llm-foundry/pull/805/files/96b6330b8c3b51294ef041f7c3110c48ca655e46#diff-df5fd070e28295f023dcec5a8ad79df3cfd90c6d7a141f9714164f8ceff0fc93R192)
which will throw error

Reply:
not sure that works with this setup...
we set a default act_fn [here](https://github.com/mosaicml/llm-foundry/pull/805/files/96b6330b8c3b51294ef041f7c3110c48ca655e46#diff-df5fd070e28295f023dcec5a8ad79df3cfd90c6d7a141f9714164f8ceff0fc93R101) in case someone uses MPTMLP on its own without using build_ffn

we could set a default [here](https://github.com/mosaicml/llm-foundry/pull/805/files/96b6330b8c3b51294ef041f7c3110c48ca655e46#diff-df5fd070e28295f023dcec5a8ad79df3cfd90c6d7a141f9714164f8ceff0fc93R177) but all ffn types allow us to specify custom act fns
(expacially ones we don't own like transformer engine ffn)

Reply:
ah 😅 defaults with python dicts are something, your suggestion above sounds good to me.

Alternatively you could explicitly handle the case below. 
```
act_fn = _DEFAULT_ACT_FN if None else resolve_ffn_act_fn(ffn_act_fn)
```